provider/file/file.go

```diff
@@ -18,68 +20,138 @@ var _ provider.Provider = (*Provider)(nil)
 // Provider holds configurations of the provider.
 type Provider struct {
 	provider.BaseProvider `mapstructure:",squash"`
+	Directory             string `description:"Load configuration from one or more .toml files in a directory"`
 }
 
 // Provide allows the file provider to provide configurations to traefik
 // using the given configuration channel.
 func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
-	watcher, err := fsnotify.NewWatcher()
+	configuration, err := p.loadConfig()
+
 	if err != nil {
-		log.Error("Error creating file watcher", err)
 		return err
 	}
 
-	file, err := os.Open(p.Filename)
+	if p.Watch {
+		var watchDir string
+
+		if p.Directory != "" {
+			watchDir = p.Directory
+		} else {
+			watchDir = filepath.Dir(p.Filename)
```

Comment:
@rjshep Of course, but when user specifies `FileName`, the watcher is added on its parent directory (`filepath.Dir(p.Filename)`). It was the mechanism before your PR.

However, before the modification the callback checked if the events received by the watcher were sent by the configuration file as described below : 

```go
case event := <-watcher.Events:
    if strings.Contains(event.Name, file.Name())
```
It would have been better to add directly the watcher on the file but, thanks to this code, the configuration was reloaded only when the `Filename` was modified.
But now, the configuration will be reloaded if one of all the toml file contained in the same directory than the `Filename`. 
There will have no problem because only the `Filename` configuration will be reloaded but, it should be better to take advantage of you PR to improve the `Filename` watching by replacing `watchDir = filepath.Dir(p.Filename)` by `watchDir = p.Filename`.

Then, the configuration will only be reloaded when the `Filename` will change.